### ISSUE TITLE
False Positive WhiteSpace.ScopeIndent error with HEREDOC

### ISSUE THREAD

**Issue by DavidAllewell** [squizlabs/PHP_CodeSniffer#3274]
**Describe the bug**
When using a HEREDOC to assign a value to a variable within a function definition, control structures within the function definition are reported to be indented too far, however, their bodies have correct indentation.

**Code sample**
```php
<?php

function function_name()
{
    if (true) {
        ;
    }

    for ($i = 1; $i < 5; $i++) {
        ;
    }

    $comment = <<<EOF
        heredoc line 1
        heredoc line 2
        EOF;
}
```

**Custom ruleset**
No - Applies to default PSR2 coding standard

**To reproduce**
1. Create a file called `test.php` with the code sample above...
2. Run `phpcs --standard=PSR2 test.php`
3. See error message displayed
```
FILE: test.php
------------------------------------------------------------------------
FOUND 4 ERRORS AFFECTING 4 LINES
------------------------------------------------------------------------
  5 | ERROR | [x] Line indented incorrectly; expected 0 spaces, found 4
  7 | ERROR | [x] Line indented incorrectly; expected 0 spaces, found 4
  9 | ERROR | [x] Line indented incorrectly; expected 0 spaces, found 4
 11 | ERROR | [x] Line indented incorrectly; expected 0 spaces, found 4
------------------------------------------------------------------------
PHPCBF CAN FIX THE 4 MARKED SNIFF VIOLATIONS AUTOMATICALLY
------------------------------------------------------------------------
```

**Expected behavior**
Correct indentation detected for control statements within function declaration.

**Versions (please complete the following information):**
 - OS: CentOS 7
 - PHP: 7.3
 - PHPCS: 3.5.8
 - Standard: PSR2
 
 **Additional Details**
 Replacing the HEREDOC with string concatenation, such as in the below code snippet, the file passes PSR2 standard
 ```
$comment = "heredoc line 1\n"
          . "heredoc line 2";

 ```

**Comment by gsherwood** [comment]
I can't replicate this on PHPCS 3.5.7, 3.5.8, or master:
```
$ cat temp.php
<?php

function function_name()
{
    if (true) {
        ;
    }

    for ($i = 1; $i < 5; $i++) {
        ;
    }

    $comment = <<<EOF
        heredoc line 1
        heredoc line 2
        EOF;
}
$ phpcs --version
PHP_CodeSniffer version 3.5.8 (stable) by Squiz (http://www.squiz.net)
$ phpcs temp.php --standard=PSR2 -v
Registering sniffs in the PSR2 standard... DONE (43 sniffs registered)
Creating file list... DONE (1 files in queue)
Changing into directory .
Processing temp.php [PHP => 70 tokens in 17 lines]... DONE in 1ms (0 errors, 0 warnings)
```
What does the verbose output show you?

**Comment by DavidAllewell** [comment]
Verbose gives me an almost identical output, however my file reports only 67 tokens, compared to your file's 70:
```
Registering sniffs in the PSR2 standard... DONE (43 sniffs registered)
Creating file list... DONE (1 files in queue)
Changing into directory .
Processing test.php [PHP => 67 tokens in 17 lines]... DONE in 6ms (4 errors, 0 warnings)
```

I've attached an output from `-vv` which will hopefully help further: [verbose.txt](https://github.com/squizlabs/PHP_CodeSniffer/files/6160596/verbose.txt)

**Comment by gsherwood** [comment]
Thanks for attaching that, this looks like the problem:
```
Process token [53]: T_START_HEREDOC => <<<EOF\n
* failed to find the end of the here/nowdoc
* token 53 changed from T_START_HEREDOC to T_STRING
```
I'll keep trying to replicate.

**Comment by gsherwood** [comment]
The code is pretty simple - it just looks for a `T_END_HEREDOC` token, which PHP assigns. It seems like your code is not being tokenized in that way.

For the code you provided, PHP should tokenize it like this (see https://3v4l.org/gM1So):
```
0 : T_OPEN_TAG => <?php\n
1 : T_WHITESPACE => \n
2 : T_FUNCTION => function
...
52 : T_WHITESPACE =>  
53 : T_START_HEREDOC => <<<EOF\n
54 : T_ENCAPSED_AND_WHITESPACE =>         heredoc line 1\n        heredoc line 2\n
55 : T_END_HEREDOC =>         EOF
56 : ;
57 : T_WHITESPACE => \n
58 : }
59 : T_WHITESPACE => \n
```

For PHP versions older than 7.3, you'd get this:
```
0 : T_OPEN_TAG => <?php\n
1 : T_WHITESPACE => \n
2 : T_FUNCTION => function
...
52 : T_WHITESPACE =>  
53 : T_START_HEREDOC => <<<EOF\n
54 : T_ENCAPSED_AND_WHITESPACE =>         heredoc line 1\n        heredoc line 2\n        EOF;\n}\n
```

Are you sure you are running PHPCS on 7.3? It looks more like you're on 7.2 or older.

**Comment by DavidAllewell** [comment]
I did wonder that, but I tried setting the version manually earlier with no change. The path to PHP 7.3 is also included in my `$PATH` variable:

```
% phpcs --config-show
Using config file: /usr/share/pear-data/PHP_CodeSniffer/CodeSniffer.conf

php_version: 70327
% php --version
PHP 7.3.27 (cli) (built: Mar  4 2021 16:00:24) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.27, Copyright (c) 1998-2018 Zend Technologies
    with the ionCube PHP Loader + ionCube24 v10.4.5, Copyright (c) 2002-2020, by ionCube Ltd.
    with Zend OPcache v7.3.27, Copyright (c) 1999-2018, by Zend Technologies
```

**Comment by gsherwood** [comment]
Then I'm out of ideas. Maybe try running the code from https://3v4l.org/gM1So as a script on your machine and making sure it tokenizes correctly.

**Comment by DavidAllewell** [comment]
Thanks for the script. I ran this and it was all tokenised correctly. I got talking with a work colleague and I believe the issue is in how it was installed.

The root user on my (and my colleague's) server defaults to a PHP version of PHP 5.4.16. I believe that, somehow, this version of PHP was being inherited when calling `phpcs` (which was installed to `/usr/bin/phpcs`). I tried reinstalling and running manually via the `.phar` file and this doesn't flag an error. I'll see if I can work out how to get around the root user's PHP version being used and I'll update this ticket with anything that might be useful.

Thanks for your help!

**Comment by gsherwood** [comment]
Glad you figured it out. Thanks for posting the outcome.
